airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java

```diff
@@ -156,21 +170,83 @@ public void runInternal(final ITransaction transaction, final IntegrationConfig
 
   @VisibleForTesting
   static void consumeWriteStream(final AirbyteMessageConsumer consumer) throws Exception {
+    consumeWriteStream(consumer, false,
+        INTERRUPT_THREAD_DELAY_MINUTES, TimeUnit.MINUTES,
+        EXIT_THREAD_DELAY_MINUTES, TimeUnit.MINUTES);
+  }
+
+  @VisibleForTesting
+  static void consumeWriteStream(final AirbyteMessageConsumer consumer,
+                                 final boolean testRun,
+                                 final int interruptTimeDelay,
+                                 final TimeUnit interruptTimeUnit,
+                                 final int exitTimeDelay,
+                                 final TimeUnit exitTimeUnit)
+      throws Exception {
     // use a Scanner that only processes new line characters to strictly abide with the
     // https://jsonlines.org/ standard
     final Scanner input = new Scanner(System.in).useDelimiter("[\r\n]+");
-    consumer.start();
-    while (input.hasNext()) {
-      final String inputString = input.next();
-      final Optional<AirbyteMessage> messageOptional = Jsons.tryDeserialize(inputString, AirbyteMessage.class);
-      if (messageOptional.isPresent()) {
-        consumer.accept(messageOptional.get());
-      } else {
-        LOGGER.error("Received invalid message: " + inputString);
+    final Thread currentThread = Thread.currentThread();
+    try {
+      consumer.start();
+      while (input.hasNext()) {
+        final String inputString = input.next();
+        final Optional<AirbyteMessage> messageOptional = Jsons.tryDeserialize(inputString, AirbyteMessage.class);
+        if (messageOptional.isPresent()) {
+          consumer.accept(messageOptional.get());
+        } else {
+          LOGGER.error("Received invalid message: " + inputString);
+        }
+      }
+    } finally {
+      final List<Thread> runningThreads = ThreadUtils.getAllThreads()
+          .stream()
+          .filter(runningThread -> !runningThread.getName().equals(currentThread.getName()) && !runningThread.isDaemon())
```

Comment:
for my own knowledge, why do we not want threads that are daemons?

Reply:
Threads marked as daemons are fine.

The ones that are not daemons are holding back the JVM process, so they can be problematic. 

That means that if you still have non-daemon threads that are still active, the JVM won't exit. If they are all daemons, the JVM will kill them and exit just fine when the main thread is done.

Reply:
got it. can you leave a comment something like
`We ignore daemon threads as the JVM automatically kills them when the main thread is done.`